framework/src/utils/PetscDMMoose.C

```diff
@@ -654,23 +687,59 @@ DMMooseGetEmbedding_Private(DM dm, IS * embedding)
               locator = displaced_problem->geomSearchData()._penetration_locators[it.first];
             }
             else
-            locator = dmm->_nl->_fe_problem.geomSearchData()._penetration_locators[it.first];
+              locator = dmm->_nl->_fe_problem.geomSearchData()._penetration_locators[it.first];
 
-            std::vector<dof_id_type> & slave_nodes = locator->_nearest_node._slave_nodes;
-            for (dof_id_type i = 0; i < slave_nodes.size(); ++i)
+            // penetration locator
+            std::map<dof_id_type, PenetrationInfo *>::iterator
+            lit  = locator->_penetration_info.begin(),
+            lend = locator->_penetration_info.end();
+            for (; lit!=lend; ++lit)
             {
-              if (locator->_has_penetrated.find(slave_nodes[i]) == locator->_has_penetrated.end())
-                continue;
-
-              // might want to use variable_first/last_local_dof instead
-              Node & slave_node = dmm->_nl->sys().get_mesh().node_ref(slave_nodes[i]);
-              dof_id_type dof = slave_node.dof_number(dmm->_nl->sys().number(), v, 0);
-              if (dof >= dofmap.first_dof() && dof < dofmap.end_dof())
-                unindices.insert(dof);
-            }
-          }
-        }
-      }
+              const dof_id_type slave_node_num = lit->first;
+              PenetrationInfo * pinfo = lit->second;
+              if (pinfo && pinfo->isCaptured())
+              {
+                Node & slave_node = dmm->_nl->sys().get_mesh().node_ref(slave_node_num);
+                dof_id_type dof = slave_node.dof_number(dmm->_nl->sys().number(), v, 0);
+                // might want to use variable_first/last_local_dof instead
+                if (dof >= dofmap.first_dof() && dof < dofmap.end_dof())
+                  unindices.insert(dof);
+                else
+                  cached_unindices.insert(dof);
+
+                std::vector<dof_id_type> evindices;
+                // indices for master element
+                evindices.clear();
+                const Elem * master_side =  pinfo->_side;
+                dofmap.dof_indices(master_side, evindices, v);
+                // indices of master sides
+                for (const auto & edof : evindices)
+                  if (edof >= dofmap.first_dof() && edof < dofmap.end_dof())
+                    unindices.insert(edof);
+                  else
+                    cached_unindices.insert(edof);
+              }// if pinfo
+            }// for penetration
+          }// for uncontact names
+        }// if there exist uncontacts
+      }// variables
```

Comment:
This whole section of code looks almost exactly the same as above. Can it be moved into a routine?